torch_em/model/unetr.py

```diff
@@ -0,0 +1,296 @@
+    Args:
+        windows (tensor): input tokens with [B * num_windows, window_size, window_size, C].
+        window_size (int): window size.
+        pad_hw (Tuple): padded height and width (Hp, Wp).
+        hw (Tuple): original height and width (H, W) before padding.
+
+    Returns:
+        x: unpartitioned sequences with [B, H, W, C].
+    """
+    Hp, Wp = pad_hw
+    H, W = hw
+    B = windows.shape[0] // (Hp * Wp // window_size // window_size)
+    x = windows.view(B, Hp // window_size, Wp // window_size, window_size, window_size, -1)
+    x = x.permute(0, 1, 3, 2, 4, 5).contiguous().view(B, Hp, Wp, -1)
+
+    if Hp > H or Wp > W:
+        x = x[:, :H, :W, :].contiguous()
+    return x
+
+
+class UNETR(nn.Module):
+    def __init__(
+        self,
+        encoder=None,
+        decoder=None,
+        out_channels=1
+    ) -> None:
+        depth = 3
+        initial_features = 64
+        gain = 2
+        features_decoder = [initial_features * gain ** i for i in range(depth + 1)][::-1]
+        scale_factors = depth * [2]
+        self.out_channels = out_channels
+
+        super().__init__()
+
+        if encoder is None:
+            self.encoder = ViTb_Sam(
+                depth=12,
+                embed_dim=768,
+                img_size=1024,
+                mlp_ratio=4,
+                norm_layer=partial(torch.nn.LayerNorm, eps=1e-6),  # type: ignore
+                num_heads=12,
+                patch_size=16,
+                qkv_bias=True,
+                use_rel_pos=True,
+                global_attn_indexes=[2, 5, 8, 11],  # type: ignore
+                window_size=14,
+                out_chans=256,
+            )
+        else:
+            self.encoder = encoder
+
+        if decoder is None:
+            self.decoder = Decoder(
+                features=features_decoder,
+                scale_factors=scale_factors[::-1],
+                conv_block_impl=ConvBlock2d,
+                sampler_impl=Upsampler2d
+            )
+        else:
+            self.decoder = decoder
+
+        self.z_inputs = ConvBlock2d(self.encoder.in_chans, features_decoder[-1])
+
+        self.base = ConvBlock2d(self.encoder.embed_dim, features_decoder[0])
+        self.out_conv = nn.Conv2d(features_decoder[-1], out_channels, 1)
+        self.final_activation = nn.Sigmoid()
+
+        self.deconv1 = Deconv2DBlock(self.encoder.embed_dim, features_decoder[0])
+        self.deconv2 = Deconv2DBlock(features_decoder[0], features_decoder[1])
+        self.deconv3 = Deconv2DBlock(features_decoder[1], features_decoder[2])
+
+        self.deconv4 = SingleDeconv2DBlock(features_decoder[-1], features_decoder[-1])
+
+        self.decoder_head = ConvBlock2d(2*features_decoder[-1], features_decoder[-1])
+
+    def preprocess(self, x: torch.Tensor) -> torch.Tensor:
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+        # TODO Why do we have hard-code image normalization here???
```

Comment:
@anwai98 Why do you use hard-coded values for the normalization here? In general why is this happening in the network, we usually do this as pre-processing independent of the network.

Reply:
I made this optional now.